src/agent/src/image.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright (c) 2021 Alibaba Cloud
+// Copyright (c) 2021, 2023 IBM Corporation
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use safe_path::scoped_join;
+use std::collections::HashMap;
+use std::env;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use anyhow::{anyhow, bail, Context, Result};
+use image_rs::image::ImageClient;
+use tokio::sync::Mutex;
+
+use crate::rpc::CONTAINER_BASE;
+use crate::AGENT_CONFIG;
+
+// A marker to merge container spec for images pulled inside guest.
+const ANNO_K8S_IMAGE_NAME: &str = "io.kubernetes.cri.image-name";
+const KATA_IMAGE_WORK_DIR: &str = "/run/image/";
+const CONFIG_JSON: &str = "config.json";
+const KATA_PAUSE_BUNDLE: &str = "/pause_bundle";
+
+const K8S_CONTAINER_TYPE_KEYS: [&str; 2] = [
+    "io.kubernetes.cri.container-type",
+    "io.kubernetes.cri-o.ContainerType",
+];
+
+#[rustfmt::skip]
+lazy_static! {
+    pub static ref IMAGE_SERVICE: Mutex<Option<ImageService>> = Mutex::new(None);
+}
+
+// Convenience function to obtain the scope logger.
+fn sl() -> slog::Logger {
+    slog_scope::logger().new(o!("subsystem" => "image"))
+}
+
+#[derive(Clone)]
+pub struct ImageService {
+    image_client: Arc<Mutex<ImageClient>>,
+    images: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl ImageService {
+    pub fn new() -> Self {
+        Self {
+            image_client: Arc::new(Mutex::new(ImageClient::new(PathBuf::from(
+                KATA_IMAGE_WORK_DIR,
+            )))),
+            images: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get the singleton instance of image service.
+    pub async fn singleton() -> Result<ImageService> {
+        IMAGE_SERVICE
+            .lock()
+            .await
+            .clone()
+            .ok_or_else(|| anyhow!("image service is uninitialized"))
+    }
+
+    async fn add_image(&self, image: String, cid: String) {
+        self.images.lock().await.insert(image, cid);
+    }
+
+    /// Set proxy environment from AGENT_CONFIG
+    fn set_proxy_env_vars() {
+        if env::var("HTTPS_PROXY").is_err() {
+            let https_proxy = &AGENT_CONFIG.https_proxy;
+            if !https_proxy.is_empty() {
+                env::set_var("HTTPS_PROXY", https_proxy);
+            }
+        }
+        if env::var("NO_PROXY").is_err() {
+            let no_proxy = &AGENT_CONFIG.no_proxy;
+            if !no_proxy.is_empty() {
+                env::set_var("NO_PROXY", no_proxy);
+            }
+        }
+    }
+
+    /// pause image is packaged in rootfs
+    fn unpack_pause_image(cid: &str, target_subpath: &str) -> Result<String> {
+        let cc_pause_bundle = Path::new(KATA_PAUSE_BUNDLE);
+        if !cc_pause_bundle.exists() {
+            bail!("Pause image not present in rootfs");
+        }
+
+        info!(sl(), "use guest pause image cid {:?}", cid);
+        let pause_bundle = Path::new(CONTAINER_BASE).join(cid).join(target_subpath);
```

Comment:
If I understand the flow fully (which I probably don't) then pull_image (which passes the cid here) is called in the image_pull_handler and the cid there comes from the add_devices call that is passed from do_create_container in rpcs.rs which calls `kata_sys_util::validate::verify_id(&cid)?;` at the start, so I think we are covered.

Reply:
@stevenhorsman - Thanks for checking the flow. I'd still prefer we call `verify_id()` here since a future change may change how/where this is called. We're manipulating files as the root user so we have to verify that param imho.

Reply:
Done, thanks for the suggesion @jodh-intel, and thanks for checking it @stevenhorsman!

Reply:
Thanks, @fidencio ! :smile: